### PR TITLE
Add release version to validators

### PR DESCRIPTION
**Pull request by pdp2121**
## High Level Overview of Change

Track validator server version by analyzing validation. 

### Context of Change

rippled version 1.8+ includes rippled version numbers in validation messages. Track these versions in the VHS API.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

## Before / After

Validators relation did not have server version before. Changes were made to include this new field.

Example API call after the change:
`/network/validator/nHUvzia57LRXr9zqnYpyFUFeKvis2tqn4DkXBVGSppt5M4nNq43C`

Example result:
<code>{
    "validation_public_key": "nHUvzia57LRXr9zqnYpyFUFeKvis2tqn4DkXBVGSppt5M4nNq43C",
    "signing_key": "n9KY4JZY11ndNbg55dThnoQdU9dii5q3egzoESVXw4Z7hu3maCba",
    "master_key": "nHUvzia57LRXr9zqnYpyFUFeKvis2tqn4DkXBVGSppt5M4nNq43C",
    "domain": "digifin.uk",
    "chain": "main",
    <b>"server_version": "1.9.2"</b>,
    "current_index": 73739542,
    "agreement_1h": {
        "missed": 1,
        "total": 930,
        "score": "0.99892",
        "incomplete": true
    },
    "agreement_24h": {
        "missed": 1,
        "total": 930,
        "score": "0.99892",
        "incomplete": true
    },
    "agreement_30day": {
        "missed": 1,
        "total": 930,
        "score": "0.99892",
        "incomplete": true
    },
    "partial": false,
    "unl": "vl.ripple.com",
    "revoked": false,
    "result": "success"
}</code>

## Test Plan

- Server version is inserted for validators table.
- Correct version is updated for the server_version column.
- Correct release (basic, rc, beta, etc.)
- server_version column will not be updated if there's no server version in the ledger.

### CODE REVIEW

File: src/connection-manager/agreement.ts
Comment (ckniffen) [inline src/connection-manager/agreement.ts:2]: You can disable multiple rules on the same line by comma separating the rules.

File: src/connection-manager/agreement.ts
Comment (mvadari) [inline src/connection-manager/agreement.ts:252]:
```suggestion
      const release_version = preceedFlagLedger(validation.ledger_index) && decodeReleaseVersion(validation.server_version)
```

File: src/connection-manager/agreement.ts
Comment (mvadari) [inline src/connection-manager/agreement.ts:261]:
```suggestion
        version: release_version,
```

File: src/shared/database/setup.ts
Comment (mvadari) [inline src/shared/database/setup.ts:124]: Might be more specific to call this `rippled_version` or something

**Comment by mvadari** [comment]
Please fill out the PR template.

File: src/connection-manager/agreement.ts
Comment (pdp2121) [reply src/connection-manager/agreement.ts:261]: I did this so that version will not be updated if there's ledger with no validation. With the change version: undefined will replace an actual version. Any ideas?

Comment (pdp2121) [reply src/connection-manager/agreement.ts]: Made some changes. Let me know what you think. Thanks Mayukha!

File: src/connection-manager/agreement.ts
Comment (natenichols) [inline src/connection-manager/agreement.ts:1]: Disable this for the specific lines, not for the file. Then reviewers can see exactly why this is being disabled.

Also, if google search for the error, you can get an example of allowed and disallowed code to fix linter errors

File: src/connection-manager/agreement.ts
Comment (natenichols) [inline src/connection-manager/agreement.ts:149]: This should be pretty easy to test if you export this function, or put it in the `shared` or `utils` and import here. 

Either way, testing this function will help prevent any versions show up incorrectly.

File: src/connection-manager/agreement.ts
Comment (natenichols) [inline src/connection-manager/agreement.ts:138]: nit: just 

```js
return `${prefix}${releaseType}${releaseNum}`
```

File: src/connection-manager/agreement.ts
Comment (natenichols) [inline src/connection-manager/agreement.ts:161]: I wouldn't throw, prefer to just return `null`

and then check for null as shown above.

File: src/shared/database/setup.ts
Comment (natenichols) [inline src/shared/database/setup.ts:124]: We'll need to carefully deploy this to staging to see if this will correctly add this column. 

If not we'll have to manually go and add the column

File: src/connection-manager/agreement.ts
Comment (natenichols) [inline src/connection-manager/agreement.ts:151]: I would probably prefer to return null, and handle the case where this version is null w/ 
```js
const decodedVersion = decodeServerVersion(...)
if (decodedVersion != null)
    await saveVersion(validator, decodedVersion)
```

File: src/connection-manager/agreement.ts
Comment (natenichols) [inline src/connection-manager/agreement.ts:260]: This seems a little complicated... 

What type are we trying to do here. Should this be type `string | null`?

**Review by natenichols — CHANGES_REQUESTED** [review]
Requesting fixes for to the uncaught throws and disabled linter rules.

File: src/connection-manager/agreement.ts
Comment (natenichols) [reply src/connection-manager/agreement.ts:260]: If it's type string or null, just pass in `decodedVersion`

**Review by ckniffen — APPROVED** [review]

File: src/connection-manager/agreement.ts
Comment (natenichols) [inline src/connection-manager/agreement.ts:117]: nit, would prefer if this was `preceedsFlagLedger()`

File: src/connection-manager/agreement.ts
Comment (natenichols) [inline src/connection-manager/agreement.ts:200]: nit: prefer `server_version != null`

File: src/shared/database/agreement.ts
Comment (natenichols) [inline src/shared/database/agreement.ts:248]: Yea, I"m okay with this nested ternary.

**Review by natenichols — APPROVED** [review]
Looks great, nice work @pdp2121.

Left a couple of nits, but nothing that should hold up approval.

**Comment by natenichols** [comment]
have we deployed this to staging yet?

File: src/connection-manager/agreement.ts
Comment (ckniffen) [reply src/connection-manager/agreement.ts:200]: That is equal to what is there.  Did you instead mean `server_version !== null`?

Comment (natenichols) [reply src/connection-manager/agreement.ts:200]: I prefer `!=`. Willing to hear other opinions, though.

`server_version != null` will not enter the `if ()` statement if we ever hit a case where `server_version` is `undefined`. 

On the off chance we hit an `undefined` server_version, we don't want to enter that if statement.

The following script:
```js
const foo = undefined;

if (foo !== null)
  console.log(`hit case !== with ${foo}`)

if (foo != null)
  console.log(`hit case != null with ${foo}`)
```

Outputs: 
```
hit case !== with undefined
```

Worth noting, I don't think this actually will change anything with the runtime behavior of this function. I _think_ `if (...)` evaluates a very similar way.

Comment (natenichols) [reply src/connection-manager/agreement.ts:200]: Also worth noting that I no longer own this repo, so if there's a style guide y'all use you're free to disregard my opinions :)

Comment (ckniffen) [reply src/connection-manager/agreement.ts:200]: Correct.  Most linting configs frown on `!=`.  I have not looked into this repos rules yet.

**Comment by natenichols** [comment]
Also, is the plan to add the `verison` field to the API in another PR?

File: src/connection-manager/agreement.ts
Comment (mvadari) [reply src/connection-manager/agreement.ts:117]: This should involve some verb like `is` or `does` - maybe `isPreceedingFlagLedger`? To make it clear that it's returning a boolean

File: src/connection-manager/agreement.ts
Comment (mvadari) [reply src/connection-manager/agreement.ts:200]: I also prefer `server_version != null` - specifically with `null` you want to use just `==` or `!=` (I believe most linters respect that, we definitely have a bunch of those in xrpl.js). That way, it also catches `undefined` and similar null-like values.

File: src/shared/database/agreement.ts
Comment (mvadari) [inline src/shared/database/agreement.ts:234]: The paranthetical here should be in line 230 as well, IMO - I was a little confused as to what this method was doing at first, and the documentation should be clear about that.

Comment (mvadari) [reply src/shared/database/agreement.ts:234]: It actually looks like this documentation is incorrect - this implies that it's only returning the type, but it's also returning the version number.

File: src/shared/database/setup.ts
Comment (mvadari) [reply src/shared/database/setup.ts:124]: This won't add the column, unless the table doesn't exist (obviously it already does in staging and prod). This is how to modify a column if the table already exists: https://github.com/ripple/validator-history-service/blob/main/src/shared/database/setup.ts#L142-#L145 (you could probably figure out how to modify this code to get it to add a column instead)

**Review by mvadari — CHANGES_REQUESTED** [review]
a few little changes (nothing major), but they should really be addressed before merging

File: src/connection-manager/agreement.ts
Comment (pdp2121) [reply src/connection-manager/agreement.ts:200]: Since `server_version = isPreceedingFlagLedger(validation.ledger_index) && decodeServerVersion(validation.server_version)`, it can also take a value of `false` when `isPreceedingFlagLedger(validation.ledger_index)` return `false`. So should it be `server_version != null && server_version !== false`?

File: src/shared/database/agreement.ts
Comment (pdp2121) [reply src/shared/database/agreement.ts:234]: It returns rc/beta number (1-63), e.g rc7, not the full version number. Should I rename the variable `releaseNum` and update the doc to avoid confusion?

File: src/shared/database/setup.ts
Comment (pdp2121) [reply src/shared/database/setup.ts:124]: I added the column in the if statement on line 131.

File: src/connection-manager/agreement.ts
Comment (natenichols) [reply src/connection-manager/agreement.ts:200]: If that's the case, I'm okay with `if (server_version)`

File: src/shared/database/agreement.ts
Comment (pdp2121) [reply src/shared/database/agreement.ts:234]: @natenichols

Comment (natenichols) [reply src/shared/database/agreement.ts:234]: I would say something like 

"Extracts the server version type from the beta version byte. This function will return nothing if a release version, `-b\<beta version\>` if a beta version, and `-rc<\release candidate version\>` if a release candidate."

instead of 

"Classify server version type"

Comment (natenichols) [reply src/shared/database/agreement.ts:234]: I actually prefer `extractServerVersionType`, because this helper does not extract the server_version, it only extracts the release type and version (i.e. nothing if release, -b1 if beta, -rc1 if release candidate).

But that's just my opinion, and am more than willing to discuss better names.

Comment (natenichols) [reply src/shared/database/agreement.ts:234]: maybe `extractReleaseType`?

File: src/api/routes/v1/validator.ts
Comment (natenichols) [inline src/api/routes/v1/validator.ts:72]: this can be 
```ts
server_version?: string
```

File: src/api/routes/v1/validator.ts
Comment (natenichols) [inline src/api/routes/v1/validator.ts:22]: same

**Review by natenichols — APPROVED** [review]

**Review by mvadari — APPROVED** [review]

**Pull request merged**
